crates/oxc_linter/src/rules/eslint/dot_notation.rs

```diff
@@ -0,0 +1,343 @@
+    ///
+    /// ### Examples
+    ///
+    /// Examples of **incorrect** code for this rule:
+    /// ```js
+    /// const x = foo["bar"];
+    ///
+    /// // ** For sample config { "allowKeywords": false } **
+    /// const foo = { "class": "CS 101" }
+    /// const x = foo["class"]; // Property name is a reserved word, square-bracket notation
+    ///                         // required
+    ///
+    /// class C {
+    ///     #in;
+    ///     foo() {
+    ///         this.#in; // Dot notation is required for private identifiers
+    ///     }
+    /// }
+    ///
+    /// // ** For sample config { "allowPattern": "^[a-z]+(_[a-z]+)+$" } **
+    /// const data = {};
+    /// data["fooBar"] = 42;
+    /// ```
+    ///
+    /// Examples of **correct** code for this rule:
+    /// ```js
+    /// const x = foo.bar;
+    /// const y = foo[bar];
+    ///
+    /// // For sample config { "allowPattern": "^[a-z]+(_[a-z]+)+$" }
+    /// const data = {};
+    /// data["foo_bar"] = 42;
+    /// ```
+    DotNotation,
+    eslint,
+    style,
+    // Fixes are possible
+    // See <https://github.com/eslint/eslint/blob/main/lib/rules/dot-notation.js>
+    pending
+);
+
+/// List of ES3 keywords.
+const KEYWORDS: &[&str] = &[
+    "abstract",
+    "boolean",
+    "break",
+    "byte",
+    "case",
+    "catch",
+    "char",
+    "class",
+    "const",
+    "continue",
+    "debugger",
+    "default",
+    "delete",
+    "do",
+    "double",
+    "else",
+    "enum",
+    "export",
+    "extends",
+    "false",
+    "final",
+    "finally",
+    "float",
+    "for",
+    "function",
+    "goto",
+    "if",
+    "implements",
+    "import",
+    "in",
+    "instanceof",
+    "int",
+    "interface",
+    "long",
+    "native",
+    "new",
+    "null",
+    "package",
+    "private",
+    "protected",
+    "public",
+    "return",
+    "short",
+    "static",
+    "super",
+    "switch",
+    "synchronized",
+    "this",
+    "throw",
+    "throws",
+    "transient",
+    "true",
+    "try",
+    "typeof",
+    "var",
+    "void",
+    "volatile",
+    "while",
+    "with",
+];
+lazy_static! {
+    static ref VALID_IDENTIFER: Regex = Regex::new(r"^[a-zA-Z_$][a-zA-Z0-9_$]*$").unwrap();
+}
+
+impl Rule for DotNotation {
+    fn from_configuration(value: serde_json::Value) -> Self {
+        let mut cfg = DotNotationConfig::default();
+
+        if let Some(config) = value.get(0) {
+            if let Some(val) = config.get("allowPattern").and_then(serde_json::Value::as_str) {
+                cfg.allow_pattern = RegexBuilder::new(val).build().ok();
+            }
+            if let Some(val) = config.get("allowKeywords").and_then(serde_json::Value::as_bool) {
+                cfg.allow_keywords = val;
+            }
+        }
+
+        Self(Box::new(cfg))
+    }
+
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        if let AstKind::MemberExpression(node) = &node.kind() {
```

Comment:
Use an early return here instead, to remove a level of indentation:

```suggestion
        let AstKind::MemberExpression(node) = &node.kind() else { return };
```